### DISCUSSION TITLE
asincio Future event loops

### DISCUSSION THREAD

**Discussion by pwelby-TFS** [deepgram/community#726] [General help] [answered]
### Which Deepgram product feature are you using?

Deepgram SDKs

### Details

I am using Cursor.sh as my IDE. I am working on creating a JARVIS clone using Python, Groq and Deepgram-SDK. I asked two previous questions and received a proper response, but on the third question to the Python application, I received an error. I then asked the internal GPT to tell me the error and this is the response: "The error message you're seeing is related to the use of asyncio and it's saying that a Future was attached to a different event loop than the one it's being awaited on. This usually happens when you're trying to use asyncio in a multi-threaded environment, or when you're not properly managing your event loops.
In your case, the error is coming from the Deepgram library's Microphone class. It seems like the Microphone class is trying to use asyncio.run_coroutine_threadsafe to run a coroutine on a different thread, but it's not properly managing the event loop.
Unfortunately, this is an issue with the Deepgram library itself, and not something you can fix directly in your code. You should report this issue to the maintainers of the Deepgram library".

### If you are making a request to the Deepgram API, what is the full Deepgram URL you are making a request to?

DEEPGRAM_URL = f"https://api.deepgram.com/v1/speak?model={self.MODEL_NAME}&performance=some&encoding=linear16&sample_rate=24000"

### If possible, please attach an example audio file to reproduce the issue.

Any .wav file

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by davidvonthenen** [comment, marked as answer]
Hi @pwelby-TFS 

Aside from what chat GPT is telling you, what problem are you actually encountering? Can you paste the error in your application here?

Granted we could do a little more error handling and checking/confirmation as they related to ensuring the loop is running and correctly (it should be), but other than that, the thread/loop should be ok.

The reason why you see the `if inspect.iscoroutinefunction(self.push_callback_org):` is because the Microphone class handles both asyncio and thread type models in a single class.
